tc/core/polyhedral/cuda/mapped_scop.cc

```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
+      domain1, domain2);
+  mappedDependences = mappedDependences.intersect(dependences);
+
+  auto space = mappedDependences.get_space();
+  auto sameThreadConstraint = getSameWarpConstraints(space, "_1", "_2", 1, numThreads);
+  auto sameWarpConstraints =
+    getSameWarpConstraints(space, "_1", "_2", 32, numThreads);
+
+  if (mappedDependences == mappedDependences.intersect_params(sameThreadConstraint)) {
+    return 0;
+  } else if (mappedDependences == mappedDependences.intersect_params(sameWarpConstraints)) {
+    return 1;
+  }
+  return 2;
+}
+
+void MappedScop::insertBestSyncInSeq(detail::ScheduleTree* seq) {
+  CHECK(seq->elemAs<detail::ScheduleTreeElemSequence>());
+
+  auto children = seq->children();
+  auto nChildren = children.size();
+
+  auto outer = hasOuterSequentialMember(scop_->scheduleRoot(), seq);
+
+  auto range = nChildren;
+  auto size = nChildren;
+  // When there is an outer sequential member, we need to check the
+  // dependences between an iteration of the loop and the next one.
+  // The sequence is repeated twice, and we consider the whole sequence
+  // to solve the problem.
+  if (outer) {
+    range = nChildren + 1;
+    size = 2 * nChildren;
+  }
+
+  std::vector<std::vector<int>> bestSync(
+      size, std::vector<int>(range, -1));
+  // Get the synchronization needed between children[i] and children[i+k]
+  // without considering the children in between
+  for (size_t k = 0; k < range; ++k) {
+    for (size_t i = 0; i + k < size; ++i) {
+      auto ik = (i + k) % nChildren;
+      auto ii = i % nChildren;
+      if(ik == ii) {
+        bestSync[i][k] = 0;
+      } else {
+        bestSync[i][k] = findBestSync(children[ii], children[ik]);
+      }
+    }
+  }
+
+  // Get the worse synchronization needed in the sequence
+  // children[i], ..., children[i+k]
+  for (size_t k = 2; k < range; ++k) {
+    for (size_t i = 0; i + k < size; ++i) {
+      bestSync[i][k] = std::max(bestSync[i][k-1], bestSync[i][k]);
+      bestSync[i][k] = std::max(bestSync[i+1][k-1], bestSync[i][k]);
+    }
+  }
+
+  // The optimal number of block sync and thread sync needed to
+  // have the sequence children[i], ..., children[j] correctly
+  // synchronized
+  std::vector<std::vector<std::pair<int, int>>> optimalValue(
+      size, std::vector<std::pair<int, int>>(range, {-1, -1}));
+
+  // An optimal position for doing a synchronization between children[i]
+  // and children[j]. The second member indicates which synchronization
+  // should be made. This should be use recursively to get the optimal
+  // synchronization. If the second member is equal to 0, this means that
+  // no synchronization is needed.
+  std::vector<std::vector<std::pair<int, int>>> optimalSyncPosition(
+      size, std::vector<std::pair<int, int>>(range, {-1, -1}));
+  for (size_t i = 0; i < size; ++i) {
+    optimalValue[i][0] = {0, 0};
+    optimalSyncPosition[i][0] = {0, 0};
+  }
+  for (size_t k = 1; k < range; ++k) {
```

Comment:
I wonder if it would be possible to extract the DP into a separate function (does not seem to need anything other than the input matrix) and write a test for it comparing against a brute-force implementation on a small size?